File: dsa/concepts/13_graphs/java/ShortestPath_L743_NetworkDelayTime.java

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Queue;

public class ShortestPath_L743_NetworkDelayTime {

	public static void main(String[] args) {
		ShortestPath_L743_NetworkDelayTime obj = new ShortestPath_L743_NetworkDelayTime();

		int[][] times = new int[][]{
            {2,1,1},
            {2,3,1},
            {3,4,1}};
		System.out.println(obj.networkDelayTime(times, 4, 2));
	}

    // Problem: https://leetcode.com/problems/network-delay-time/
    // Idea: We need to find the shartest path to each vertex. Max of all these is the result.
	public int networkDelayTime(int[][] times, int n, int k) {
        HashMap<Integer, ArrayList<int[]>> adj = new HashMap<>();
        for(int i=1; i<=n; i++)
            adj.put(i, new ArrayList<>());
        for(int[] edge: times){
            int s = edge[0], d = edge[1], t = edge[2];
            adj.get(s).add(new int[]{t, d});
        }

        int[] dist = new int[n+1];      // not needed
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[k] = 0;

        HashSet<Integer> vis = new HashSet<>();
        Queue<int[]> q = new PriorityQueue<>((a, b) -> a[0]-b[0]);
        q.add(new int[]{0, k});
        int maxTime = 0;
        while(!q.isEmpty()){
            int[] curr = q.remove();
            int time = curr[0], node = curr[1];
            if(vis.contains(node))
                continue;

            vis.add(node);
            maxTime = Math.max(maxTime, time);
            for(int[] nei: adj.get(node)){
                int neiNode = nei[1], edgeTime = nei[0];
                if(!vis.contains(neiNode)){
                    if(time+edgeTime < dist[neiNode]){
                        dist[neiNode] = time+edgeTime;
                        q.add(new int[]{dist[neiNode], neiNode});
                    }
                }
            }
        }

        return vis.size() == n ? maxTime: -1;
    }
}

File: dsa/concepts/13_graphs/python/shortest_path_l743_network_delay_time.py
from typing import List
from collections import defaultdict
import heapq


"""
Problem: https://leetcode.com/problems/network-delay-time/
Idea:
We need to find the shortest time to reach every node from source k.
This is a standard Dijkstra problem.
The answer is the maximum distance among all reachable nodes.
If any node is unreachable, return -1.

Time Complexity: O(E log V)
Space Complexity: O(V + E)
"""
def network_delay_time(times: List[List[int]], n: int, k: int) -> int:
    adj = defaultdict(list)
    for src, dest, time in times:
        adj[src].append((time, dest))

    dist = [float('inf')] * (n + 1)
    dist[k] = 0

    visited = set()
    pq = []
    heapq.heappush(pq, (0, k))   # (time, node)

    max_time = 0

    while pq:
        time, node = heapq.heappop(pq)

        if node in visited:
            continue

        visited.add(node)
        max_time = max(max_time, time)

        for edge_time, nei in adj[node]:
            if nei not in visited and time + edge_time < dist[nei]:
                dist[nei] = time + edge_time
                heapq.heappush(pq, (dist[nei], nei))

    return max_time if len(visited) == n else -1


# one test (same as Java main)
if __name__ == "__main__":
    times = [
        [2, 1, 1],
        [2, 3, 1],
        [3, 4, 1]
    ]
    print(network_delay_time(times, 4, 2))  # 2
